### PR TITLE
[TensorIR][UX] Type annotation-based runtime type checking

### PR DESCRIPTION
**Pull request by junrushao1994**
Addressed the UX issue raised in #9551.

To address the first issue (cryptic error message on packed function boundary), this PR introduced a decorator @type_checked to do runtime type checking and generate more user-friendly error messages. In our example:

image

The second issue is mainly wording. This PR rephrasing the error message to:

```
ValueError: The loop no long exists in the IRModule
```

However, I cannot find any environment variable to disable stacktrace printing. It was there months ago but no longer exists...Weird...

The notebook: https://gist.github.com/junrushao1994/cd0a33dc966da3b3f1abfa50a14e6d2c

CC: @tqchen

### CODE REVIEW

**Review by Hzfengsy — APPROVED** [review]

**Comment by junrushao1994** [comment]
Looks like we need at least python 3.7 to run the decorator…

**Comment by tqchen** [comment]
on the c++ stacktrace side, perhaps the main thing is to update the trace formatter to simplify and skip some of the less informative traces(e.g. unpackcall_dispatcher) but keep the rest of the useful traces

File: src/tir/schedule/concrete_schedule.h [hunk 1]
```diff
@@ -204,7 +204,7 @@ inline StmtSRef ConcreteScheduleNode::GetSRef(const BlockRV& block_rv) const {
                << (obj.defined() ? obj->GetTypeKey() : "None");
   }
   if (sref->stmt == nullptr) {
-    LOG(FATAL) << "ValueError: The StmtSRef has expired";
+    LOG(FATAL) << "ValueError: The block no long exists in the IRModule";
```
Comment (tqchen) [inline src/tir/schedule/concrete_schedule.h:207]: no longer

**Comment by junrushao1994** [comment]
> on the c++ stacktrace side, perhaps the main thing is to update the [trace normalizer](https://github.com/apache/tvm/blob/main/src/runtime/c_runtime_api.cc#L225) to simplify and skip some of the less informative traces(e.g. unpackcall_dispatcher) but keep the rest of the useful traces

Alternatively, there are other metaprogramming tricks to completely avoid recursion and could potentially generate significantly better error message.

File: src/tir/schedule/concrete_schedule.h (hunk 1 above)
Comment (junrushao1994) [reply src/tir/schedule/concrete_schedule.h:207]: Ooops typo

**Comment by junrushao1994** [comment]
Update with a py36 compatible implementation

**Comment by junrushao1994** [comment]
The error message is improved to:

```
TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_307968/1279557970.py in <module>
      5 i = sch.get_loops(block_c)
      6 # Tile the loop nesting.
----> 7 i_0, i_1, i_2 = sch.split(i, factors=[None, 4, 4])
      8 print(sch.mod.script())

~/Projects/tvm-dev/python/tvm/tir/schedule/_type_checker.py in wrap(*args, **kwargs)
    147                 )
    148                 if error_msg is not None:
--> 149                     raise TypeError(error_msg)
    150         return func(*args, **kwargs)
    151 

TypeError: In "Schedule.split", "loop" has wrong type. Expected "LoopRV", but gets: "list"
```

**Review by tqchen — APPROVED** [review]

**Pull request merged**